# Numpy

In [ ]:
import numpy as np

**Array Creation**

In [ ]:
# Create an array from a list
arr = np.array([1, 2, 3, 4])
print("Array:", arr)

Array: [1 2 3 4]


In [ ]:
# Create an array filled with zeros
zeros_array = np.zeros((2, 3))
print("Zeros Array:\n", zeros_array)

Zeros Array:
 [[0. 0. 0.]
 [0. 0. 0.]]


In [ ]:
# Create an array filled with ones
ones_array = np.ones((2, 3))
print("Ones Array:\n", ones_array)


Ones Array:
 [[1. 1. 1.]
 [1. 1. 1.]]


In [ ]:
# Create an uninitialized array
empty_array = np.empty((2, 3))
print("Empty Array:\n", empty_array)

Empty Array:
 [[1. 1. 1.]
 [1. 1. 1.]]


In [ ]:
# Create an array with a range of values
range_array = np.arange(5)
print("Range Array:", range_array)

Range Array: [0 1 2 3 4]


In [ ]:
# Create an array with evenly spaced values
linspace_array = np.linspace(0, 1, 5)
print("Linspace Array:", linspace_array)

Linspace Array: [0.   0.25 0.5  0.75 1.  ]


In [ ]:
# Create an array with values spaced evenly on a log scale
logspace_array = np.logspace(0, 2, 3)
print("Logspace Array:", logspace_array)

Logspace Array: [  1.  10. 100.]


In [ ]:
# Create coordinate matrices from coordinate vectors
x = np.array([1, 2, 3])
y = np.array([4, 5])
X, Y = np.meshgrid(x, y)
print("X:\n", X)
print("Y:\n", Y)

# Stack coordinates into pairs
print("Coordinates List")
coords = np.array([X.ravel(), Y.ravel()]).T
print(coords)

#or in tuples
print("Coordinates tuples")
coordinate_pairs = list(zip(X.ravel().tolist(), Y.ravel().tolist()))
print(coordinate_pairs)

X:
 [[1 2 3]
 [1 2 3]]
Y:
 [[4 4 4]
 [5 5 5]]
Coordinates List
[[1 4]
 [2 4]
 [3 4]
 [1 5]
 [2 5]
 [3 5]]
Coordinates tuples
[(1, 4), (2, 4), (3, 4), (1, 5), (2, 5), (3, 5)]


In [ ]:
# Create an identity matrix
identity_matrix = np.eye(3)
print("Identity Matrix:\n", identity_matrix)

Identity Matrix:
 [[1. 0. 0.]
 [0. 1. 0.]
 [0. 0. 1.]]


In [ ]:
# Create a diagonal matrix
diag_matrix = np.diag([1, 2, 3])
print("Diagonal Matrix:\n", diag_matrix)

Diagonal Matrix:
 [[1 0 0]
 [0 2 0]
 [0 0 3]]


**List Vs Array**

In [ ]:
# Python list
lst = [1, 2, 3, 4]
lst_squared = [x**2 for x in lst]
print("List squared:", lst_squared)

# NumPy array
arr = np.array([1, 2, 3, 4])
arr_squared = arr**2
print("Array squared:", arr_squared)

List squared: [1, 4, 9, 16]
Array squared: [ 1  4  9 16]


**Array Manipulation**

In [ ]:
# Reshape an array
arr = np.arange(6)
reshaped_array = arr.reshape((2, 3))
print("Reshaped Array:\n", reshaped_array)

Reshaped Array:
 [[0 1 2]
 [3 4 5]]


In [ ]:
# Flatten an array
flattened_array = reshaped_array.ravel()
print("Flattened Array:", flattened_array)

Flattened Array: [0 1 2 3 4 5]


In [ ]:
# Transpose an array
transposed_array = reshaped_array.T
print("Transposed Array:\n", transposed_array)

Transposed Array:
 [[0 3]
 [1 4]
 [2 5]]


In [ ]:
# Swap two axes of an array
swapped_axes_array = np.swapaxes(reshaped_array, 0, 1)
print("Swapped Axes Array:\n", swapped_axes_array)

Swapped Axes Array:
 [[0 3]
 [1 4]
 [2 5]]


In [ ]:
# Concatenate arrays
arr1 = np.array([1, 2, 3])
arr2 = np.array([4, 5, 6])
concatenated_array = np.concatenate((arr1, arr2))
print("Concatenated Array:", concatenated_array)

Concatenated Array: [1 2 3 4 5 6]


In [ ]:
# Split an array into multiple sub-arrays
split_array = np.split(concatenated_array, 2)
print("Split Array:", split_array)

Split Array: [array([1, 2, 3]), array([4, 5, 6])]


In [ ]:
# Stack arrays horizontally
hstack_array = np.hstack((arr1, arr2))
print("Horizontal Stack Array:", hstack_array)

Horizontal Stack Array: [1 2 3 4 5 6]


In [ ]:
a1=np.array([[1,2,3],[4,5,6]])
a2=np.array([7,8])
print(a1.shape)
print(a2.shape)
horizontal_stacking=np.hstack((a1,a2.reshape(2,1)))
print(horizontal_stacking)
print(horizontal_stacking.shape)


(2, 3)
(2,)
[[1 2 3 7]
 [4 5 6 8]]
(2, 4)


In [ ]:
# Stack arrays vertically
vstack_array = np.vstack((arr1, arr2))
print("Vertical Stack Array:\n", vstack_array)

Vertical Stack Array:
 [[1 2 3]
 [4 5 6]]


In [ ]:
# Repeat elements of an array
repeated_array = np.repeat(arr1, 2)
print("Repeated Array:", repeated_array)

Repeated Array: [1 1 2 2 3 3]


In [ ]:
# Repeat array in a 2D grid
tiled_array_2d = np.tile([[1, 2], [3, 4]], (2, 3))
print("Tiled Array 2D:\n", tiled_array_2d)


Tiled Array 2D:
 [[1 2 1 2 1 2]
 [3 4 3 4 3 4]
 [1 2 1 2 1 2]
 [3 4 3 4 3 4]]


**Array Indexing**

In [ ]:
matrix = np.array([[10, 20, 30],
                   [40, 50, 60],
                   [70, 80, 90]])


In [ ]:
matrix.shape

(3, 3)

In [ ]:
# Accessing elements
print("Element at (0,1):", matrix[0, 1])     # row 0, col 1 → 20
print("First row:", matrix[0, :])           # all columns of first row
print("Second column:", matrix[:, 1])       # all rows, col 1
print("Sub-matrix (first 2 rows, last 2 cols):", matrix[:2, 1:])

Element at (0,1): 20
First row: [10 20 30]
Second column: [20 50 80]
Sub-matrix (first 2 rows, last 2 cols): [[20 30]
 [50 60]]


In [ ]:
matrix[1, 2] = 100
print("Modified matrix:\n", matrix)

matrix[:, 0] = [1, 2, 3]   # replace first column
print("After replacing first column:\n", matrix)

Modified matrix:
 [[ 10  20  30]
 [ 40  50 100]
 [ 70  80  90]]
After replacing first column:
 [[  1  20  30]
 [  2  50 100]
 [  3  80  90]]


**Broadcasting**

When performing operations between arrays:

Compare shapes element by element from right to left.

Dimensions must be either:

Equal, OR

One of them is 1 (so it can be stretched).

If neither condition holds → error.

In [ ]:
arr = np.array([10, 20, 30])
result = arr + 5
print(result)

[15 25 35]


In [ ]:
a = np.array([[1], [2], [3]])
b = np.array([10, 20, 30])

result = a + b
print(result)

[[11 21 31]
 [12 22 32]
 [13 23 33]]


In [ ]:
#example
sales = np.array([[100, 200, 300],
                  [400, 500, 600],
                  [700, 800, 900]])

discount = np.array([10, 20, 30])  # shape (3,)

# Broadcasts discount across each row
new_sales = sales - discount
print(new_sales)

[[ 90 180 270]
 [390 480 570]
 [690 780 870]]


**Mathematical Operation**s

In [ ]:
# Add arrays element-wise
arr_sum = np.add([1, 2, 3], [4, 5, 6])
print("Sum:", arr_sum)

Sum: [5 7 9]


In [ ]:
# Subtract arrays element-wise
arr_diff = np.subtract([4, 5, 6], [1, 2, 3])
print("Difference:", arr_diff)

Difference: [3 3 3]


In [ ]:
# Multiply arrays element-wise
arr_prod = np.multiply([1, 2, 3], [4, 5, 6])
print("Product:", arr_prod)

Product: [ 4 10 18]


In [ ]:
# Divide arrays element-wise
arr_div = np.divide([4, 5, 6], [1, 2, 3])
print("Division:", arr_div)

Division: [4.  2.5 2. ]


In [ ]:
# Raise elements to the specified power
arr_pow = np.power([1, 2, 3], 2)
print("Power:", arr_pow)

Power: [1 4 9]


In [ ]:
# Compute the square root of each element
arr_sqrt = np.sqrt([1, 4, 9])
print("Square Root:", arr_sqrt)

Square Root: [1. 2. 3.]


In [ ]:
# Compute the exponential of each element
arr_exp = np.exp([1, 2, 3])
print("Exponential:", arr_exp)

Exponential: [ 2.71828183  7.3890561  20.08553692]


In [ ]:
# Compute the natural logarithm of each element
arr_log = np.log([1, np.e, np.e**2])
print("Logarithm:", arr_log)

Logarithm: [0. 1. 2.]


In [ ]:
# Compute trigonometric sine of each element
arr_sin = np.sin([0, np.pi/2, np.pi])
print("Sine:", arr_sin)

Sine: [0.0000000e+00 1.0000000e+00 1.2246468e-16]


In [ ]:
# Compute trigonometric cosine of each element
arr_cos = np.cos([0, np.pi/2, np.pi])
print("Cosine:", arr_cos)

Cosine: [ 1.000000e+00  6.123234e-17 -1.000000e+00]


**Statistical Operations**

In [ ]:
# Compute the mean
mean_value = np.mean([1, 2, 3, 4, 5])
print("Mean:", mean_value)


# Compute the median
median_value = np.median([1, 2, 3, 4, 5])
print("Median:", median_value)

# Compute the standard deviation
std_deviation = np.std([1, 2, 3, 4, 5])
print("Standard Deviation:", std_deviation)


# Compute the variance
variance = np.var([1, 2, 3, 4, 5])
print("Variance:", variance)


# Compute the nth percentile
percentile_50 = np.percentile([1, 2, 3, 4, 5], 50)
print("50th Percentile:", percentile_50)



Mean: 3.0
Median: 3.0
Standard Deviation: 1.4142135623730951
Variance: 2.0
50th Percentile: 3.0


**Random Number Generation**

In [ ]:
# Generate a single random number
random_number = np.random.rand()
print("Random Number:", random_number)

# Generate an array of random numbers with shape (2, 3)
random_array = np.random.rand(2, 3)
print("Random Array:\n", random_array)


Random Number: 0.4219527047666932
Random Array:
 [[0.31903931 0.66582113 0.33813624]
 [0.24977323 0.60931441 0.24521858]]


In [ ]:
# Generate a single random integer from 0 to 10 (exclusive)
random_int = np.random.randint(0, 10)
print("Random Integer:", random_int)

# Generate an array of random integers with shape (2, 3) from 0 to 10 (exclusive)
random_array = np.random.randint(0, 10, size=(2, 3))
print("Random Integer Array:\n", random_array)


Random Integer: 1
Random Integer Array:
 [[9 6 7]
 [9 4 5]]


In [ ]:
# Generate a single random sample from an array
array = [1, 2, 3, 4, 5]
random_sample = np.random.choice(array)
print("Random Sample:", random_sample)


# Generate an array of 3 random samples from the array with replacement
random_samples = np.random.choice(array, size=3, replace=True)
print("Random Samples with Replacement:", random_samples)


# Generate an array of 3 random samples from the array without replacement
random_samples_no_replace = np.random.choice(array, size=3, replace=False)
print("Random Samples without Replacement:", random_samples_no_replace)


Random Sample: 1
Random Samples with Replacement: [3 2 4]
Random Samples without Replacement: [5 2 1]


In [ ]:
# Generate a single random number from a uniform distribution between 1 and 10
random_number = np.random.uniform(1, 10)
print("Random Number from Uniform Distribution:", random_number)

# Generate an array of random numbers with shape (2, 3) from a uniform distribution
random_array = np.random.uniform(1, 10, size=(2, 3))
print("Random Array from Uniform Distribution:\n", random_array)

Random Number from Uniform Distribution: 3.428790409302218
Random Array from Uniform Distribution:
 [[7.37238291 8.80294604 9.13130422]
 [1.23087996 5.83471316 9.85048098]]


In [ ]:
# Generate a single random number from a normal distribution with mean=0 and std=1
random_number = np.random.normal(loc=0, scale=1)
print("Random Number from Normal Distribution:", random_number)

# Generate an array of random numbers with shape (2, 3) from a normal distribution
random_array = np.random.normal(loc=0, scale=1, size=(2, 3))
print("Random Array from Normal Distribution:\n", random_array)

Random Number from Normal Distribution: 0.01334132098989265
Random Array from Normal Distribution:
 [[ 1.22987527  1.16635549 -1.61460712]
 [ 2.82736792 -0.2331169   0.53221459]]


In [ ]:
# Generate a single random number from an exponential distribution with scale=1
random_number = np.random.exponential(scale=1)
print("Random Number from Exponential Distribution:", random_number)

# Generate an array of random numbers with shape (2, 3) from an exponential distribution
random_array = np.random.exponential(scale=1, size=(2, 3))
print("Random Array from Exponential Distribution:\n", random_array)

Random Number from Exponential Distribution: 1.610220713249646
Random Array from Exponential Distribution:
 [[1.39813528 0.0505943  2.50709568]
 [0.75315337 0.76345891 0.80361949]]


In [ ]:
# Generate a single random number from a binomial distribution
random_number = np.random.binomial(n=10, p=0.5)
print("Random Number from Binomial Distribution:", random_number)


# Generate an array of random numbers with shape (2, 3) from a binomial distribution
random_array = np.random.binomial(n=10, p=0.5, size=(2, 3))
print("Random Array from Binomial Distribution:\n", random_array)

Random Number from Binomial Distribution: 6
Random Array from Binomial Distribution:
 [[7 7 5]
 [4 4 4]]


In [ ]:
# Generate a single random number from a Poisson distribution with lambda=3
random_number = np.random.poisson(lam=3)
print("Random Number from Poisson Distribution:", random_number)

# Generate an array of random numbers with shape (2, 3) from a Poisson distribution
random_array = np.random.poisson(lam=3, size=(2, 3))
print("Random Array from Poisson Distribution:\n", random_array)

Random Number from Poisson Distribution: 2
Random Array from Poisson Distribution:
 [[4 3 5]
 [1 2 1]]


In [ ]:
# Seed the random number generator
np.random.seed(42)

# Generate a random number
random_number = np.random.rand()
print("Random Number with Seed:", random_number)

# Generate an array of random numbers with shape (2, 3) with the same seed
random_array = np.random.rand(2, 3)
print("Random Array with Seed:\n", random_array)

Random Number with Seed: 0.3745401188473625
Random Array with Seed:
 [[0.95071431 0.73199394 0.59865848]
 [0.15601864 0.15599452 0.05808361]]


**Linear algebra**

In [ ]:
import numpy as np

# Create a matrix from a list of lists
matrix = np.array([[1, 2], [3, 4]])
print("Matrix:\n", matrix)

Matrix:
 [[1 2]
 [3 4]]


In [ ]:
# Transpose a matrix
transposed_matrix = matrix.T
print("Transposed Matrix:\n", transposed_matrix)

Transposed Matrix:
 [[1 3]
 [2 4]]


In [ ]:
# Matrix multiplication
matrix1 = np.array([[1, 2], [3, 4]])
matrix2 = np.array([[2, 0], [1, 3]])
product_matrix = np.dot(matrix1, matrix2)
print("Matrix Product:\n", product_matrix)

Matrix Product:
 [[ 4  6]
 [10 12]]


In [ ]:
elementwise_product = matrix1 * matrix2
print("Element-wise Product:\n", elementwise_product)

Element-wise Product:
 [[ 2  0]
 [ 3 12]]


In [ ]:
# Compute the inverse of a matrix
inverse_matrix = np.linalg.inv(matrix)
print("Inverse Matrix:\n", inverse_matrix)

Inverse Matrix:
 [[-2.   1. ]
 [ 1.5 -0.5]]


In [ ]:
# Compute the determinant of a matrix
determinant = np.linalg.det(matrix)
print("Determinant:", determinant)

Determinant: -2.0000000000000004


In [ ]:
# Compute eigenvalues and eigenvectors
eigenvalues, eigenvectors = np.linalg.eig(matrix)
print("Eigenvalues:", eigenvalues)
print("Eigenvectors:\n", eigenvectors)

Eigenvalues: [-0.37228132  5.37228132]
Eigenvectors:
 [[-0.82456484 -0.41597356]
 [ 0.56576746 -0.90937671]]


In [ ]:
# Coefficient matrix
A = np.array([[2, 3], [4, -1]])

# Right-hand side vector
b = np.array([13, 3])

# Solve the system of linear equations
solution = np.linalg.solve(A, b)
print("Solution (x, y):", solution)

Solution (x, y): [1.57142857 3.28571429]


**Calculus**

In [ ]:
import numpy as np

# Define a function
def f(x):
    return x**2 + 2*x + 1

# Define a point at which to compute the derivative
x = 2.0

# Define a small step size
h = 1e-5

# Compute the derivative using the finite difference method
df_dx = (f(x + h) - f(x - h)) / (2 * h)
print("Numerical Derivative:", df_dx)

Numerical Derivative: 5.9999999999504885


**Statistical Analayis using NumPy**

In [ ]:
import numpy as np

# Sample data
data = [5, 7, 8, 5, 9, 10, 6, 7, 8, 9]

# Convert to a numpy array for convenience
data_array = np.array(data)

# Mean
mean = np.mean(data_array)

# Median
median = np.median(data_array)

# Mode (Note: numpy doesn't have a mode function, so we'll use np.unique for a simple approach)
values, counts = np.unique(data_array, return_counts=True)
mode = values[np.argmax(counts)]

# Range
data_range = np.max(data_array) - np.min(data_array)

# Variance
variance = np.var(data_array, ddof=0)  # Population variance

# Standard Deviation
std_dev = np.std(data_array, ddof=0)  # Population standard deviation

# Interquartile Range (IQR)
q75, q25 = np.percentile(data_array, [75, 25])
iqr = q75 - q25

# Skewness
mean_diff = data_array - mean
skewness = np.mean(mean_diff**3) / (np.std(data_array, ddof=0)**3)

# Kurtosis
kurtosis = np.mean(mean_diff**4) / (np.std(data_array, ddof=0)**4) - 3

# Print results
print(f"Mean: {mean}")
print(f"Median: {median}")
print(f"Mode: {mode}")
print(f"Range: {data_range}")
print(f"Variance: {variance}")
print(f"Standard Deviation: {std_dev}")
print(f"IQR: {iqr}")
print(f"Skewness: {skewness}")
print(f"Kurtosis: {kurtosis}")


Mean: 7.4
Median: 7.5
Mode: 5
Range: 5
Variance: 2.6399999999999997
Standard Deviation: 1.624807680927192
IQR: 2.5
Skewness: -0.1007112198921821
Kurtosis: -1.1446280991735536


# Assignment

**Task 1.1:** Create a NumPy array representing hourly temperature readings for a day.

  *) Find min, max, and average temperature.

  *) Convert Celsius readings to Fahrenheit.


**Task 1.2:** Explain the difference between Python Lists and Numpy arrays in terms of memory and operations.

1. Memory: NumPy arrays are more memory efficient since they store data of the same type
   in contiguous memory blocks, unlike Python lists which are object-based.
2. Operations: NumPy arrays support vectorized operations (fast computations without loops),
   whereas Python lists need explicit iteration.

**Task 2.1:** Create an array of 1000 random customer ages (between 18 and 60).Find the percentage of customers in the age group 25–35.


**Task 3.1:** Given daily stock prices for a company in a NumPy array:

      prices = np.array([120, 121, 119, 122, 118, 117, 115, 116, 118, 120])

          *) Extract prices for days 2–6.

          *)  Get all prices greater than 118.

**Task 4.1:**
Create a 3×3 matrix of employee salaries in thousands.Add a bonus vector [5, 10, 15] to each row using broadcasting.


**Task 5.1:** You have monthly sales for 3 years in a 1D array (length 36).

            *) Reshape into a 3×12 matrix (3 years, 12 months).
            *) Find total sales per year.
            *) Find the month with the highest average sales across years.